nullaway/src/test/java/com/uber/nullaway/tools/SerializationTestHelper.java

```diff
@@ -0,0 +1,137 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.tools;
+
+import static org.junit.Assert.fail;
+
+import com.google.errorprone.CompilationTestHelper;
+import com.uber.nullaway.NullAway;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class SerializationTestHelper {
+  private final List<FixDisplay> fixDisplays = new ArrayList<>();
+  private final Path outputPath;
+  private CompilationTestHelper compilationTestHelper;
+
+  public SerializationTestHelper(Path outputPath) {
+    this.outputPath = outputPath.resolve("fixes.tsv");
+  }
+
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  public SerializationTestHelper addSourceLines(String path, String... lines) {
+    compilationTestHelper.addSourceLines(path, lines);
+    return this;
+  }
+
+  public SerializationTestHelper addFixes(FixDisplay... fixDisplays) {
+    this.fixDisplays.addAll(Arrays.asList(fixDisplays));
+    return this;
+  }
+
+  public SerializationTestHelper setArgs(List<String> args) {
+    compilationTestHelper =
+        CompilationTestHelper.newInstance(NullAway.class, getClass()).setArgs(args);
+    return this;
+  }
+
+  public void doTest() {
+    clearOutput();
+    try {
+      compilationTestHelper.doTest();
+    } catch (Exception e) {
+      System.out.println("Ignored");
+    }
+    FixDisplay[] outputFixDisplays;
+    outputFixDisplays = readOutputFixes();
+    compareFixes(outputFixDisplays);
+  }
+
+  private void clearOutput() {
+    try {
+      Files.deleteIfExists(outputPath);
+    } catch (IOException ignored) {
+      throw new RuntimeException("Did not expect it");
+    }
+  }
+
+  private void compareFixes(FixDisplay[] outputFixDisplays) {
+    ArrayList<FixDisplay> notFound = new ArrayList<>();
+    ArrayList<FixDisplay> output = new ArrayList<>(Arrays.asList(outputFixDisplays));
```

Comment:
7117c411de9d53661f7c80f814cf5c892a1ddbc3

These lists contains duplicate fixes, for every fix at ```expectedOutputFixes```, we check if it is available at the output and if not found, it is added to the list of ```notFound```.

If we expect NullAway to find the list of ```f0, f0, f0, f1, f1, f1```, all five fixes must be found by NullAway and will fail the test otherwise.

Reply:
That doesn't quite answer my question. My comment here is why are we copying around and changing the data type of `outputFixDisplays` back and forth. Specifically:

1) It gets created as an `ArrayList` in `readOutputFixes()`, then turned to an array upon return from that function
2) Then in `doTest()` it is passed between the result of `readOutputFixes()` and the argument to `compareFixes(...)`
3) The first thing `compareFixes(...)` does is turn it back into an `ArrayList` (via copying)

Is the copying being done for the sake of immutability at the method boundaries? I guess that makes enough sense to me in test code, where we care more about correctness than performance overhead. It just feels a bit weird.